sources/RuleRunner.cpp

```diff
@@ -1329,6 +1330,96 @@ ActionHasTarget(int8 type)
 }
 
 
+static bool
+getDirectoryPath(BString& str, const entry_ref& ref)
+{
+	BEntry entry(&ref);
+	if (entry.InitCheck() != B_OK)
+		return false;
+
+	if (!entry.IsDirectory() && entry.GetParent(&entry) != B_OK)
+		return false;
+
+	BPath path(&entry);
+	if (path.InitCheck() != B_OK)
+		return false;
+
+	str = path.Path();
+	return true;
+}
+
+
+bool
+SetTextForType(BString& text, int8 type, const entry_ref& ref, bool isTest)
+{
+	if (isTest)
+		switch (type) {
+			case TEST_TYPE:
+			{
+				BNode node(&ref);
+				if (node.InitCheck() != B_OK)
+					return false;
+
+				BNodeInfo nodeInfo(&node);
+				if (nodeInfo.InitCheck() != B_OK)
+					return false;
+
+				char mimeType[B_MIME_TYPE_LENGTH];
+				if (nodeInfo.GetType(mimeType) != B_OK)
+					return false;
+
+				text = mimeType;
+				return true;
+			}
+			case TEST_NAME:
+				text = ref.name;
+				return true;
+			case TEST_SIZE:
+			{
+				BEntry entry(&ref);
+				if (entry.InitCheck() != B_OK || !entry.IsFile())
+					return false;
+
+				off_t size;
+				if (entry.GetSize(&size) != B_OK)
+					return false;
+
+				text = "";
+				text << size;
+				return true;
+			}
+			case TEST_LOCATION:
+				return getDirectoryPath(text, ref);
+			default:
+				return false;
+		}
+	else
+		switch (type) {
+			case ACTION_MOVE:
+			case ACTION_COPY:
+				return getDirectoryPath(text, ref);
+			case ACTION_RENAME:
+				text = ref.name;
+				return true;
+			case ACTION_ARCHIVE:
+			{
+				BEntry entry(&ref);
+				if (entry.InitCheck() != B_OK || !entry.IsFile())
+					return false;
+
+				BPath path(&ref);
+				if (path.InitCheck() != B_OK)
+					return false;
+
+				text = path.Path();
+				return true;
```

Comment:
I think it'd be best to behave like Haiku's [Zip-O-Matic](http://cgit.haiku-os.org/haiku/tree/src/add-ons/tracker/zipomatic) Tracker add-on.
- If a file/folder "one-file" is zipped, the archive is named "one-file.zip"
- If several files/folders are zipped, the archive is called "Archive.zip"

I Filer doesn't need to copy this Zip-O-Matic behaviour:
- If that archive name is already taken, Zip-O-Matic names it "Archive 2.zip" or "one-file 2.zip" (increasing numbers)

We want to do exactly what the ACTION says, "Add to archive". Which "zip" does by default.
(What do you say to changing the zip parameters to "zip -9ury", i.e. add "-y to store symbolic links as the link instead of the referenced file"? Sounds useful to me.

The "Archive" part is localized, so something like this:
```
BString archive(B_TRANSLATE_COMMENT("Archive", "Default name of zip archives (same as Zip-O-Matic)"));
```
A `archive =+ ".zip";` isn't even needed, as your man page research turned out.